# Manisha Pednekar NLP Homework 6

<a id='BacktoTop'></a>
## <font color=blue> Contents </font>
- [Question 1: Evaluate text similarity of Amazon book search results ](#section_ID1)
- [Question 2: Evaluate capsule strings with search phrase using google](#section_ID2)

In [1]:
from contractions import CONTRACTION_MAP

In [2]:
import nltk
import re
import pattern
from pattern.en import parsetree


In [5]:
from normalization import normalize_corpus
import numpy as np

In [14]:
def compute_cosine_similarity(doc_features, corpus_features, top_n = 23):

    # get document vectors
    doc_features = doc_features.toarray()[0]
    corpus_features = corpus_features.toarray()

    # compute similarities
    similarity = np.dot(doc_features, 
                        corpus_features.T)

    # get docs with highest similarity scores
    top_docs = similarity.argsort()[::-1][:top_n]
    
    top_docs_with_score = [(index, round(similarity[index], 3))
                            for index in top_docs]
    
    return top_docs_with_score

### <font color=red> 1. Evaluate text similarity of Amazon book search results by doing the following: </font>
 - a.	Do a book search on Amazon. Manually copy the full book title (including subtitle) of each of the top 24 books listed in the first two pages of search results. 
 - b.	In Python, run one of the text-similarity measures covered in this course, e.g., cosine similarity. Compare each of the book titles, pairwise, to every other one. 

### I searched the book with the phrase **"emperor of all maladies"** in Amazon book search interface. The complete book name is **"The Emperor of All Maladies: A Biography of Cancer"**

In [6]:
#--------All thr first 24 searches from Amazon book search for phrase "emperor of all maladies"
toy_corpus = ["The Emperor of All Maladies: A Biography of Cancer",
"Sapiens: A Brief History of Humankind",
"Being Mortal: Medicine and What Matters in the End",
"Interpreter of Maladies",
"The Laws of Medicine: Field Notes from an Uncertain Science (TED Books)",
"The Gene: An Intimate History",
"When Breath Becomes Air",
"A Brief History of Everyone Who Ever Lived: The Human Story Retold Through Our Genes",
"A Necessary Evil: A Novel (Wyndham & Banerjee Series)",
"Smoke and Ashes: A Novel (Wyndham & Banerjee Series)",
"Homo Deus: A Brief History of Tomorrow",
"A Rising Man: A Novel (Wyndham & Banerjee Series)",
"El gen / The Gene: An Intimate History: Una historia personal (Spanish Edition)",
"A State of Freedom: A Novel",
"The Emperor of All Maladies: A Biography of Cancer (Chinese Edition)",
"Healing Justice: Stories of Wisdom and Love (How to Die Smiling Book 3)",
"TED Books Box Set: The Completist: The Terrorist's Son, The Mathematics of Love, The Art of Stillness, The Future of Architecture, Beyond Measure, ... The Laws of Medicine, and Follow Your Gut",
"The Best American Science and Nature Writing 2013 (The Best American Series)",
"Summary & Study Guide - The Emperor of All Maladies: A Biography of Cancer",
"TED Books Box Set: The Science Mind: Follow Your Gut, How We'll Live on Mars, and The Laws of Medicine",
"Summary - The Gene: Book by Siddhartha Mukherjee - An Intimate History (The Gene: An Intimate History - Book, Paperback, Hardcover, Summary, Audiobook, Audible.",
"Siddhartha Mukherjee: The Laws of Medicine : Field Notes from an Uncertain Science (Hardcover); 2015 Edition",
"A Cancer in the Family: Take Control of Your Genetic Inheritance",
"The Emperor of All Maladies by Siddhartha Mukherjee - Key Takeaways & Analysis: A Biography of Cancer"]

[Back to the top](#BacktoTop)

In [19]:
# code from Dipanjan sarkar's book chapter 6 source code:
#https://github.com/dipanjanS/text-analytics-with-python/blob/master/Old-First-Edition/source_code/Ch06_Text_Similarity_and_Clustering/document_similarity.py
from utils import build_feature_matrix
query_docs = toy_corpus
# normalize and extract features from the toy corpus
norm_corpus = normalize_corpus(toy_corpus, lemmatize=True)
tfidf_vectorizer, tfidf_features = build_feature_matrix(norm_corpus,
                                                        feature_type='tfidf',
                                                        ngram_range=(1, 1), 
                                                        min_df=0.0, max_df=1.0)
                                                        
# normalize and extract features from the query corpus
norm_query_docs =  normalize_corpus(query_docs, lemmatize=True)            
query_docs_tfidf = tfidf_vectorizer.transform(norm_query_docs)

def compute_cosine_similarity(doc_features, corpus_features,
                              top_n=23):
    # get document vectors
    doc_features = doc_features.toarray()[0]
    corpus_features = corpus_features.toarray()
    # compute similarities
    similarity = np.dot(doc_features, 
                        corpus_features.T)
    # get docs with highest similarity scores
    top_docs = similarity.argsort()[::-1][:top_n]
    top_docs_with_score = [(index, round(similarity[index], 3))
                            for index in top_docs]
    return top_docs_with_score

    
print ('Document Similarity Analysis using Cosine Similarity')
print ('='*60)
print("\n")
for index, doc in enumerate(query_docs):
    
    doc_tfidf = query_docs_tfidf[index]
    top_similar_docs = compute_cosine_similarity(doc_tfidf,
                                             tfidf_features,
                                             top_n=23)
    top_similar_docs.pop(0)
    highDocNum = top_similar_docs[0][0]
    lowDocNum = top_similar_docs[-1][0]
    print( 'Document',index+1 ,':', doc)
    #print ('Top', len(top_similar_docs), 'similar docs:')
    print ('Highest Similarity Doc num {} : {} \nHighest Similarity Score: {}'.format(highDocNum+1, toy_corpus[highDocNum], top_similar_docs[0][1]))
    print('Lowest Similarity Doc num {} : {} \nLowest Similarity Score: {}'.format(lowDocNum+1, toy_corpus[lowDocNum],top_similar_docs[-1][1]))
    
    print ('-'*80 )
    print("\n")
    

Document Similarity Analysis using Cosine Similarity


Document 1 : The Emperor of All Maladies: A Biography of Cancer
Highest Similarity Doc num 15 : The Emperor of All Maladies: A Biography of Cancer (Chinese Edition) 
Highest Similarity Score: 0.744
Lowest Similarity Doc num 18 : The Best American Science and Nature Writing 2013 (The Best American Series) 
Lowest Similarity Score: 0.0
--------------------------------------------------------------------------------


Document 2 : Sapiens: A Brief History of Humankind
Highest Similarity Doc num 11 : Homo Deus: A Brief History of Tomorrow 
Highest Similarity Score: 0.301
Lowest Similarity Doc num 22 : Siddhartha Mukherjee: The Laws of Medicine : Field Notes from an Uncertain Science (Hardcover); 2015 Edition 
Lowest Similarity Score: 0.0
--------------------------------------------------------------------------------


Document 3 : Being Mortal: Medicine and What Matters in the End
Highest Similarity Doc num 5 : The Laws of Medicine: F

<a id="section_ID1"></a>
### <font color=red> Which two titles are the most similar to each other ? Which are the most dissimilar ?</font>

- Document 1 : The Emperor of All Maladies: A Biography of Cancer
 - Highest Similarity Doc num 15 : The Emperor of All Maladies: A Biography of Cancer (Chinese Edition) 
 - Highest Similarity Score: 0.744

- The next higher similar Doc num 24: The Emperor of All Maladies by Siddhartha Mukherjee - Key Takeaways & Analysis: A Biography of Cancer
 - Similarity Score: 0.568

- Lowest Similarity Doc num 18 : The Best American Science and Nature Writing 2013 (The Best American Series) 
 - Lowest Similarity Score: 0.0


I searched the book with the phrase **"emperor of all maladies"** in Amazon book search interface. The complete book name is **"The Emperor of All Maladies: A Biography of Cancer"** and this title is matched the most with the **search number 15 "The Emperor of All Maladies: A Biography of Cancer (Chinese Edition)" with score of 0.744**. The only difference between these 2 book titles is that the two words added at the end as "(Chinese Edition)" of the exact book title selected, thus these scores seem relevant as per the words of the matched title. We used TF-IDF formula while calculating cosine similarity which make it weighs down the frequent words and I think that is the reason that the book title is more similar with this "The Emperor of All Maladies: A Biography of Cancer (Chinese Edition)" with score 0.744 than "The Emperor of All Maladies by Siddhartha Mukherjee - Key Takeaways & Analysis: A Biography of Cancer" with score 0.568. The book with **lowest cosine similarity score of 0.0 is the search number 18 with book title "The Best American Science and Nature Writing 2013 (The Best American Series)". This result of lowest cosine score of this title which have too generic words like science, nature, best with the original title seems correct knowing that the TF-IDF formula used while calculating cosine similarity.** 

[Back to the top](#BacktoTop)

<a id="section_ID2"></a>
#### <font color=red> Q2.	Now evaluate using a major search engine.  </font>
 - a.	Enter one of the book titles from question 1a into Google, Bing, or Yahoo!. Copy the capsule of the first organic result and the 20th organic result. Take web results only (i.e., not video results), and skip sponsored results. 
 - b.	Run the same text similarity calculation that you used for question 1b on each of these capsules in comparison to the original query (book title). 
 - c.	Which one has the highest similarity measure?

**Below, we compared cosine similarities between the book title and the two capsule search strings. From the visual inspection of the results, the capsule number 2 ("Cancer: The Emperor of All Maladies The AACR was...") has the most matched words from the book title with cosine similarity score of 0.537.**

In [22]:
capsule = ["The Emperor of All Maladies: A Biography of Cancer", 
           "Winner of the Pulitzer Prize, and now a documentary from Ken Burns on PBS, The Emperor of All Maladies is a magnificent, profoundly humane “biography” of cancer—from its first documented appearances thousands of years ago through the epic battles in the twentieth century to cure, control, and conquer it to a radical ...", 
           "Cancer: The Emperor of All Maladies The AACR was proud to be a production supporter of the PBS documentary CANCER: THE EMPEROR OF ALL MALADIES ..."]

query_docs = capsule

# normalize and extract features from the toy corpus
norm_corpus = normalize_corpus(capsule, lemmatize=True)
tfidf_vectorizer, tfidf_features = build_feature_matrix(norm_corpus,
                                                        feature_type='tfidf',
                                                        ngram_range=(1, 1), 
                                                        min_df=0.0, max_df=1.0)
                                                        
# normalize and extract features from the query corpus
norm_query_docs =  normalize_corpus(query_docs, lemmatize=True)            
query_docs_tfidf = tfidf_vectorizer.transform(norm_query_docs)

print ('Document Similarity Analysis in Book title search and the returned search capsules using Google')
print ('='*80)
print("\n")
for index, doc in enumerate(query_docs):
    
    doc_tfidf = query_docs_tfidf[index]
    top_similar_docs = compute_cosine_similarity(doc_tfidf,
                                             tfidf_features,
                                             top_n=2)
    top_similar_docs.pop(0)
    print( 'Document',index+1 ,':', doc)
    print ('Highest cosine similar docs:')
    print ('-'*80 )
    for doc_index, sim_score in top_similar_docs:
        print( 'Doc num: {} Similarity Score: {}\nDoc: {}'.format(doc_index+1,
                                                                 sim_score,
                                                                 capsule[doc_index]) )
        print ('-'*80 )
        print("\n")
    print            


Document Similarity Analysis in Book title search and the returned search capsules using Google


Document 1 : The Emperor of All Maladies: A Biography of Cancer
Highest cosine similar docs:
--------------------------------------------------------------------------------
Doc num: 3 Similarity Score: 0.537
Doc: Cancer: The Emperor of All Maladies The AACR was proud to be a production supporter of the PBS documentary CANCER: THE EMPEROR OF ALL MALADIES ...
--------------------------------------------------------------------------------


Document 2 : Winner of the Pulitzer Prize, and now a documentary from Ken Burns on PBS, The Emperor of All Maladies is a magnificent, profoundly humane “biography” of cancer—from its first documented appearances thousands of years ago through the epic battles in the twentieth century to cure, control, and conquer it to a radical ...
Highest cosine similar docs:
--------------------------------------------------------------------------------
Doc num: 1 Si

[Back to the top](#BacktoTop)